### PR TITLE
Add IsDelegatable and IsPageable flags to UDF class

### PR DESCRIPTION
**Pull request by rick-nguyen**
These flags will allow the binder to mark a UDF as delegatable/pageable correctly based on the rule bound to it.

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs [hunk 1]
```diff
@@ -22,28 +24,45 @@
...
         private readonly bool _isImperative;
         private readonly IEnumerable<UDFArg> _args;
         private TexlBinding _binding;
 
-        public override bool IsAsync => _binding?.IsAsync(UdfBody) ?? false;
+        public override bool IsAsync => _binding?.IsAsync(UdfBody) ?? false;
+
+        public bool IsPageable => _binding?.IsPageable(_binding.Top) ?? false;
+
+        public bool IsDelegatable => _binding?.IsDelegatable(_binding.Top) ?? false;
```
Comment (adithyaselv) [inline src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs:47]: Can you add few binding tests with simple and deeply nested expressions using this mock datasource [https://github.com/microsoft/Power-Fx/blob/c24014662051e2fc6740e4c291f26210e1bfcf37/src/tests/Microsoft.PowerFx.Core.Tests.Shared/Helpers/TestTabularDataSource.cs#L293](https://github.com/microsoft/Power-Fx/blob/c24014662051e2fc6740e4c291f26210e1bfcf37/src/tests/Microsoft.PowerFx.Core.Tests.Shared/Helpers/TestTabularDataSource.cs#L293) ?

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs [hunk 2]
```diff
@@ -22,30 +24,47 @@
...
     /// </summary>
-    internal class UserDefinedFunction : TexlFunction
+    internal class UserDefinedFunction : TexlFunction, IExternalPageableSymbol, IExternalDelegatableSymbol
     {
         private readonly bool _isImperative;
         private readonly IEnumerable<UDFArg> _args;
         private TexlBinding _binding;
 
-        public override bool IsAsync => _binding?.IsAsync(UdfBody) ?? false;
+        public override bool IsAsync => _binding?.IsAsync(UdfBody) ?? false;
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs:43]: >_binding? [](http://example.com/codeflow?start=40&length=9)

_binidng can't be null, right?

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs [hunk 3]
```diff
@@ -22,30 +24,47 @@
...
+
+        public bool IsPageable => _binding?.IsPageable(_binding.Top) ?? false;
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs:45]: > _binding? [](http://example.com/codeflow?start=33&length=10)

? not needed.

**Review by Shpakh-MSFT — DISMISSED** [review]
:shipit:

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs [hunk 4]
```diff
@@ -906,6 +906,7 @@ public virtual bool IsAsyncInvocation(CallNode callNode, TexlBinding binding)
         {
             Contracts.AssertValue(callNode);
             Contracts.AssertValue(binding);
+            Contracts.Assert(binding.GetInfo(callNode).Function is not UserDefinedFunction udf || udf.Binding != null);
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs:909]: > Contracts.Assert(binding.GetInfo(callNode).Function is not UserDefinedFunction udf || udf.Binding != null); [](http://example.com/codeflow?start=11&length=108)

This assert here doesn't make sense. You already have assert in override.

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs [hunk 5]
```diff
@@ -929,6 +930,7 @@ public virtual bool SupportsPaging(CallNode callNode, TexlBinding binding)
...
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunction.cs:933]: > Contracts.Assert(binding.GetInfo(callNode).Function is not UserDefinedFunction udf || udf.Binding != null); [](http://example.com/codeflow?start=11&length=108)

Same here.

File: src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs [hunk 6]
```diff
@@ -31,22 +34,42 @@ namespace Microsoft.PowerFx.Core.Functions
...
+
+        public bool IsPageable => _binding.IsPageable(_binding.Top);
+
+        public bool IsDelegatable => _binding.IsDelegatable(_binding.Top);
+
+        public override bool IsServerDelegatable(CallNode callNode, TexlBinding binding)
+        {
+            Contracts.AssertValue(callNode);
+            Contracts.AssertValue(binding);
+            Contracts.Assert(binding.GetInfo(callNode).Function is not UserDefinedFunction udf || udf.Binding != null);
```
Comment (Shpakh-MSFT) [inline src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs:53]: >  Contracts.Assert(binding.GetInfo(callNode).Function is not UserDefinedFunction udf || udf.Binding != null); [](http://example.com/codeflow?start=10&length=109)

This assert needs to be

  Contracts.Assert(binding.GetInfo(callNode).Function is UserDefinedFunction udf && udf.Binding != null);

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by akshar2401 — APPROVED** [review]

**Pull request merged**